### PR TITLE
[2.4] wait for dockerInfo to be set before creating node plan

### PR DESCRIPTION
**Pull request by kinarashah**
Problem:
creating node plan with empty dockerRootDir results into wrong kubelet's process

https://github.com/rancher/rancher/issues/26790

Problem:
change in node plan of registering nodes triggers cluster upgrade for other worker nodes

https://github.com/rancher/rancher/issues/26807

### CODE REVIEW

File: pkg/controllers/management/rkeworkerupgrader/plan.go [hunk 1]
```diff
@@ -30,6 +30,8 @@ func (uh *upgradeHandler) nonWorkerPlan(node *v3.Node, cluster *v3.Cluster) (*v3
 	hostAddress := node.Status.NodeConfig.Address
 	hostDockerInfo := infos[hostAddress]
 
+	logrus.Tracef("getDockerInfo for node [%s] dockerInfo %#v", node.Name, hostDockerInfo.DockerRootDir)
```
Comment (StrongMonkey) [inline pkg/controllers/management/rkeworkerupgrader/plan.go:33]: Good to know this. What does tracef do? :smiley:

File: pkg/controllers/management/rkeworkerupgrader/plan.go [hunk 2]
```diff
@@ -72,6 +74,11 @@ func (uh *upgradeHandler) workerPlan(node *v3.Node, cluster *v3.Cluster) (*v3.RK
 	hostAddress := node.Status.NodeConfig.Address
 	hostDockerInfo := infos[hostAddress]
 
+	if hostDockerInfo.Driver == "" || hostDockerInfo.DockerRootDir == "" {
```
Comment (StrongMonkey) [inline pkg/controllers/management/rkeworkerupgrader/plan.go:77]: What's the workflow in here if error is returned?

File: pkg/controllers/management/rkeworkerupgrader/plan.go (hunk 1 above)
Comment (kinarashah) [reply pkg/controllers/management/rkeworkerupgrader/plan.go:33]: we've recently started printing out whole objects in trace than debug so debug becomes more readable https://github.com/rancher/rancher/issues/24994

File: pkg/controllers/management/rkeworkerupgrader/plan.go (hunk 2 above)
Comment (kinarashah) [reply pkg/controllers/management/rkeworkerupgrader/plan.go:77]: it would return back to the sync, and keep retrying till the docker info gets assigned

**Comment by kinarashah** [comment]
@superseb @StrongMonkey 

Changing the approach a little. 
DockerInfo gets synced on the node each time agent requests for a node plan. There's no point in returning error in controllers or in `GetDockerInfo` because retry there isn't going to fix anything + spam logs. I've changed logic to wait for `Node.Status.DockerInfo` to be set on before getting and setting the node plan. 

The second commit is for issue that's happening because of this wrong node plan. First node is stuck in `registering` and on addition of other nodes, rkeConfig changes and `toUpgradeCluster` becomes true because registering node now has the right node plan (dockerInfo got synced). 

This issue shouldn't happen now, but let's say node is stuck in registering state for some other reason, any subsequent `rke up` shouldn't result into workers upgrade because of change in this node plan -- so I'm ignoring registering nodes.

**Comment by superseb** [comment]
Makes sense to me, should we add a logline for each found condition to trace down the cause of not getting a node plan?

**Comment by StrongMonkey** [comment]
Also build failed

**Pull request merged**